File: src/com/nimbusly/isometrix/Game.java
package com.nimbusly.isometrix;

import android.util.Log;

public class Game {

	private static final String TAG = "Ground";

	private Point position;
	private int width = 0;
	private int height = 0;
	private Point origin, destination;
	private long startTime;
	private int walkState;
	private int facing;
	private static final int TILE_SIZE = 32;
	
	private static final int[][] DIRECTION_INDEX = new int[][] {
		new int[] { 6, 7, 0 }, 
		new int[] { 5, 0, 1 }, 
		new int[] { 4, 3, 2 }
	};
	private long speed = 400; // ms per tile
	private long time;
	
	public void move(int dx, int dy) {
		//Log.d(TAG, "Moving " + dx + "," + dy);
		synchronized(this) {
			facing = directionIndex(dx, dy);
			position.x = (position.x + dx) % width;
			position.y = (position.y + dy) % width;
		}
	}
	
	public void updateState(long time) {
		this.time = time;
		walkState = (int) Math.abs((time / 100) % 12);
		// compute movement
		if (destination != null && !position.equals(destination)) {
			double distance = origin.distanceTo(destination);
			long step = (time - startTime) * TILE_SIZE / speed;
			if (step < distance) {
				int dx = (int) Math.round(step * (destination.x - origin.x) / distance);
				int dy = (int) Math.round(step * (destination.y - origin.y) / distance);
				facing = directionIndex(dx, dy);
				position.set(origin.x + dx, origin.y + dy);
			} else {
				position.set(destination.x, destination.y);
			}
		}
	}
	
	public Game() {
		position = new Point(width / 2, height / 2);
	}
	
	public int getWidth() {
		return width;
	}
	
	public int getHeight() {
		return height;
	}
	
	public int getXPosition() { return position.x; }
	public int getYPosition() { return position.y; }
	public Point getPosition() { return position; }
	public int getWalkState() { return walkState; }
	public int getFacing() { return facing; }
	
	public int tileAt(Point p) {
		int col = p.x / TILE_SIZE;
		int row = p.y / TILE_SIZE;
		return (col + row) % 2;
	}
	
	public int getTileSize() {
		return TILE_SIZE;
	}
	
	public void setPosition(Point point) {
		position.set(point.x, point.y);
	}
	
	public void setDestination(Point point) {
		destination = point;
		origin = new Point(position.x, position.y);
		startTime = time;
	}
	
	private int directionIndex(int dx, int dy) {
		return DIRECTION_INDEX[IntegerMath.signum(dx)+1][IntegerMath.signum(dy)+1];
	}

	public void setSpeed(long speed) {
		this.speed = speed;
	}
}

File: src/com/nimbusly/isometrix/Game.scala
package com.nimbusly.isometrix

class Game {

	private val TAG = "Ground"

	private var width = 0
	private var height = 0
	private var position = new Point(width / 2, height / 2)
	private var origin : Point = null
	private var destination : Point = null 
	private var startTime = 0L
	private var walkState = 0
	private var facing = 0
	val tileSize = 32
	
	private val DIRECTION_INDEX = Array(Array(6, 7, 0), Array(5, 0, 1), Array(4, 3, 2))
	
	private var speed = 400L // ms per tile
	private var time = 0L
	
	def move(dx : Int, dy : Int) {
		//Log.d(TAG, "Moving " + dx + "," + dy);
		this.synchronized {
			facing = directionIndex(dx, dy)
			position.x = (position.x + dx) % width
			position.y = (position.y + dy) % width
		}
	}
	
	def updateState(time : Long) {
		this.time = time
		walkState = math.abs((time / 100) % 12).asInstanceOf[Int]
		// compute movement
		if (destination != null && !position.equals(destination)) {
			val distance = origin.distanceTo(destination)
			val step = (time - startTime) * tileSize / speed
			if (step < distance) {
				val dx = step * (destination.x - origin.x) / distance
				val dy = step * (destination.y - origin.y) / distance
				facing = directionIndex(dx.asInstanceOf[Int], dy.asInstanceOf[Int])
				position.set((origin.x + dx).asInstanceOf[Int], (origin.y + dy).asInstanceOf[Int])
			} else {
				position.set(destination.x, destination.y)
			}
		}
	}
	
	def tileAt(p : Point) : Int = {
		val col = p.x / tileSize
		val row = p.y / tileSize
		(col + row) % 2
	}
	
	def setDestination(point : Point) {
		destination = point
		origin = new Point(position.x, position.y)
		startTime = time
	}
	
	private def directionIndex(dx : Int, dy : Int) : Int = {
		return DIRECTION_INDEX(IntegerMath.signum(dx)+1)(IntegerMath.signum(dy)+1)
	}
}